nexus/src/db/datastore/silo_user.rs

```diff
@@ -64,6 +67,76 @@ impl DataStore {
             })
     }
 
+    /// Delete a Silo User
+    pub async fn silo_user_delete(
+        &self,
+        opctx: &OpContext,
+        authz_silo_user: &authz::SiloUser,
+    ) -> DeleteResult {
+        opctx.authorize(authz::Action::Delete, authz_silo_user).await?;
+
+        // Delete the user and everything associated with them.
+        // TODO-scalability Many of these should be done in batches.
+        // TODO-robustness We might consider the RFD 192 "rcgen" pattern as well
+        // so that people can't, say, login while we do this.
+        self.pool_authorized(opctx)
+            .await?
+            .transaction_async(|mut conn| async move {
+                // Delete console sessions.
+                {
+                    use db::schema::console_session::dsl;
+                    diesel::delete(dsl::console_session)
+                        .filter(dsl::silo_user_id.eq(authz_silo_user.id()))
+                        .execute_async(&mut conn)
+                        .await?;
+                }
+
+                // Delete device authentication tokens.
+                {
+                    use db::schema::device_access_token::dsl;
+                    diesel::delete(dsl::device_access_token)
+                        .filter(dsl::silo_user_id.eq(authz_silo_user.id()))
+                        .execute_async(&mut conn)
+                        .await?;
+                }
+
+                // Delete group memberships.
+                {
+                    use db::schema::silo_group_membership::dsl;
+                    diesel::delete(dsl::silo_group_membership)
+                        .filter(dsl::silo_user_id.eq(authz_silo_user.id()))
+                        .execute_async(&mut conn)
+                        .await?;
+                }
+
+                // Delete ssh keys.
+                {
+                    use db::schema::ssh_key::dsl;
+                    diesel::update(dsl::ssh_key)
+                        .filter(dsl::silo_user_id.eq(authz_silo_user.id()))
+                        .filter(dsl::time_deleted.is_null())
+                        .set(dsl::time_deleted.eq(Utc::now()))
+                        .execute_async(&mut conn)
+                        .await?;
+                }
+
+                // Delete the user record.
+                // XXX-dap should use check_if_exists/execute_and_check
+                use db::schema::silo_user::dsl;
+                diesel::update(dsl::silo_user)
+                    .filter(dsl::id.eq(authz_silo_user.id()))
+                    .filter(dsl::time_deleted.is_null())
+                    .set(dsl::time_deleted.eq(Utc::now()))
+                    .execute_async(&mut conn)
+                    //.check_if_exists::<SiloUser>(authz_silo_user.id())
```

Comment:
@smklein I was hoping to use `check_if_exists()`/`execute_and_check()` here.  When I do that (i.e., comment-out L130, uncomment L131-132, and add the `use` statement for `UpdateAndCheck`), I get:

```
error[E0308]: mismatched types
   --> nexus/src/db/datastore/silo_user.rs:133:40
    |
133 |                     .execute_and_check(&mut conn)
    |                      ----------------- ^^^^^^^^^ expected struct `bb8::Pool`, found struct `async_bb8_diesel::Connection`
    |                      |
    |                      arguments to this function are incorrect
    |
    = note:      expected reference `&bb8::Pool<async_bb8_diesel::ConnectionManager<DTraceConnection<PgConnection>>>`
            found mutable reference `&mut async_bb8_diesel::Connection<DTraceConnection<PgConnection>>`
note: associated function defined here
   --> nexus/src/db/update_and_check.rs:156:18
    |
156 |     pub async fn execute_and_check(
    |                  ^^^^^^^^^^^^^^^^^
157 |         self,
158 |         pool: &bb8::Pool<ConnectionManager<DbConnection>>,
    |         -------------------------------------------------

For more information about this error, try `rustc --explain E0308`.
error: could not compile `omicron-nexus` due to previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `omicron-nexus` due to previous error
```

I guess `execute_and_check()` is not intended to be run under a transaction?  I also tried converting this transaction to use `transaction()` instead of `transaction_async()` and got a similar error:

```
error[E0308]: mismatched types
   --> nexus/src/db/datastore/silo_user.rs:127:40
    |
127 |                     .execute_and_check(conn)
    |                      ----------------- ^^^^ expected struct `bb8::Pool`, found struct `DTraceConnection`
    |                      |
    |                      arguments to this function are incorrect
    |
    = note:      expected reference `&bb8::Pool<async_bb8_diesel::ConnectionManager<DTraceConnection<PgConnection>>>`
            found mutable reference `&mut DTraceConnection<PgConnection>`
note: associated function defined here
   --> nexus/src/db/update_and_check.rs:156:18
    |
156 |     pub async fn execute_and_check(
    |                  ^^^^^^^^^^^^^^^^^
157 |         self,
158 |         pool: &bb8::Pool<ConnectionManager<DbConnection>>,
    |         -------------------------------------------------

```

For now, I could explicitly fetch the record up front and bail out if I don't find it.  Is there something else you think I should do here?

Reply:
I've fixed this in https://github.com/oxidecomputer/omicron/pull/1784/commits/be021c88ce21af7811c3e7f2d8f6a4cfcfbdbd8c by making `execute_and_check` generic over the connection. Now either pools or connections can be used.

(Feel free to revert / modify that commit as you need, but I figured code would be easier to share than a comment :P )

Reply:
Thanks for that!